api/data_refinery_api/views.py

```diff
@@ -606,34 +609,99 @@ def get(self, request, format=None):
+            'open': jobs.filter(start_time__isnull=False, end_time__isnull=True).count(),
+            # via https://stackoverflow.com/questions/32520655/get-average-of-difference-of-datetime-fields-in-django
+            'average_time': jobs.filter(start_time__isnull=False, end_time__isnull=False).aggregate(
+                average_time=Avg(F('end_time') - F('start_time')))['average_time']
+        }
+
+        if result['average_time'] is None:
+            result['average_time'] = 0
+        else:
+            result['average_time'] = result['average_time'].total_seconds()
+
+        if range_param is not None:
+            result['timeline'] = self._jobs_timeline(jobs, range_param)
+
+        return result
+
+    def _get_object_stats(self, objects, range_param):
+        result = {
+            'total': objects.count()
+        }
+
+        if range_param is not None:
+            result['timeline'] = self._created_timeline(Sample.objects, range_param)
+
+        return result
+
+    interval_timedelta = {
+        'day': dt.timedelta(days=1),
+        'week': dt.timedelta(weeks=1),
+        'month': dt.timedelta(weeks=4),
+        'year': dt.timedelta(weeks=52)
+    }
+
+    interval_timestep = {
+        'day': dt.timedelta(hours=1),
+        'week': dt.timedelta(days=1),
+        'month': dt.timedelta(days=2),
+        'year': dt.timedelta(weeks=4)
+    }
+
+    def _get_time_intervals(self, range_param):
+        current_date = dt.datetime.now(tz=timezone.utc)
+        time_step = self.interval_timestep.get(range_param)
+        start_date = current_date - self.interval_timedelta.get(range_param)
+
+        intervals = [(current_date - time_step*(i+1), current_date - time_step*i)
+                     for i in range(100) if current_date - time_step*(i+1) > start_date]
+        return intervals[::-1]
+
+    def _get_job_interval(self, jobs, start, end):
+        filtered_jobs = jobs.filter(created_at__gte=start, created_at__lte=end)
+        pending = filtered_jobs and jobs.filter(start_time__isnull=True)
+        failed = filtered_jobs and jobs.filter(success=False)
+        completed = filtered_jobs and jobs.filter(success=True)
+        open = filtered_jobs and jobs.filter(success__isnull=True)
+
+        return {
+            'start': start,
+            'end': end,
+            'total': filtered_jobs.count(),
+            'completed': completed.count(),
+            'pending': pending.count(),
+            'failed': failed.count(),
+            'open': open.count()
+        }
+
+    def _jobs_timeline(self, jobs, range_param):
+        return [self._get_job_interval(jobs, start, end) for (start, end) in self._get_time_intervals(range_param)]
+
+    def _created_timeline(self, objects, range_param):
+        return [({'start': start, 'end': end, 'total': objects.filter(created_at__gte=start, created_at__lte=end).count()}) for (start, end) in self._get_time_intervals(range_param)]
```

Comment:
This and the return line above it have bad [JonesComplexity](https://github.com/Miserlou/JonesComplexity)

Reply:
I think the simplest way to make this line more maintainable is something like

```
        results = []
        for start, end in self._get_time_intervals(range_param):
            total = objects.filter(created_at__gte=start, created_at__lte=end).count()
            stats = {   'start': start, 
                        'end': end, 
                        'total': total
                    }
            results.append(stats)

```

Reply:
Good Idea! Thanks!